src/dotty/tools/dotc/transform/TreeChecker.scala

```diff
@@ -181,6 +181,106 @@ class TreeChecker extends Phase with SymTransformer {
       if (tree.symbol.maybeOwner.isTerm)
         assert(nowDefinedSyms contains tree.symbol, i"undefined symbol ${tree.symbol}")
 
+    /** assert Java classes are not used as objects */
+    def assertIdentNotJavaClass(tree: untpd.Tree)(implicit ctx: Context): Unit = tree match {
+      case _ : untpd.Ident =>
+        assert(!tree.symbol.is(JavaModule), "Java class can't be used as value: " + tree)
+      case _ =>
+    }
+
+    /** check Java classes are not used as objects */
+    def checkIdentNotJavaClass(tree: untpd.Tree)(implicit ctx: Context): Unit = tree match {
+      // case tree: untpd.Ident =>
+      // case tree: untpd.Select =>
+      // case tree: untpd.SelectFromTypeTree =>
+      // case tree: untpd.Bind =>
+      case vd : ValDef =>
```

Comment:
I keep the unchecked cases as comments to make it explicit what cases are ignored, in order to facilitate code maintenance. I'm not sure if it's a good style.